Sources/Articles/Documentation.docc/2025/gha-faster-ci/gha-faster-ci.md

```diff
@@ -0,0 +1,547 @@
+
+Note that `actions/cache` follows some branch protection rules so it's not vulnerable to cache poisoning attacks. That means that it'll only restore a cache if it was cached from the current branch or the primary branch of your repository.
+
+After running the tests is over, you have the "Cache .build" step.
+
+That's where you're using the `save` capability of `actions/cache` to save your `.build` directory using the same `key` that you used in "Restore .build".
+
+With the `if` condition, you're avoiding spending time in the cache step when you already found an exact match for the cache key in the "Restore .build" step. This is because `actions/cache` rejects saving cache entries with duplicate keys, and does not have an update mechanism.
+
+This `if` statement will still evaluate to `false` if "Restore .build" had to fallback to `restore-keys` to find a cache.
+
+The `${{ github.event.pull_request.base.sha || github.event.after }}` in the cache `key` will also resolve to the base branch commit SHA of your pull request, or to the current commit SHA if the job run is triggered by a push event and not in a pull request.
+
+This makes sure your pull request will use a cache from the commit the pull request is based on. Push events will also usually fallback to the latest cache that has been uploaded from the branch the CI is running on.
+
+Now let's see. How is your CI doing after adding these 2 steps?
+
+For the first run, don't expect any time improvements as there is simply no cache the action run can use.
+
+But if you re-run the CI job, you'll notice a big difference.
+
+Your tests CI runtime has drop to 6 minutes 30s, saving 3 minutes 30s.
+
+This is assuming both your restore and cache steps are triggered, which is not always true thanks to the if condition you have in "Cache .build". But in a lot of CIs you'll still be restoring and uploading caches in the same CI run. For example every single time you push changes to your primary branch.
+
+![Tests CI With Cache](tests-ci-with-cache.png)
+
+You've already saved at least 3.5 minutes of CI runtime, and that's great. But let's look closer.
+
+What's consuming the CI runtime now that you're using caching? Is it really taking 6.5 minutes for the tests to run even after all your efforts?
+
+The answer is No. If you look at the "Restore .build" step, you'll notice it's taking around 30 seconds to restore the data. Likewise, "Cache .build" step of the first run of your CI is taking more than 2.5 minutes caching around 1.5 GB worth of `.build` data.
+
+Wouldn't it be so nice if you could decrease the times spent caching? After all, only 3 minutes of your CI runtime is related to your tests step.
+
+## Speed Up Caching With zstd
+
+As it turns out, `actions/cache` compresses directories before uploading them to GitHub. This compression process uses `gzip` by default. Considering `gzip` runs on a single CPU core, it can easily take 2.5 minutes to compress and upload a 1.5 GB `.build` directory.
```

Comment:
I'd add a short paragraph here about `zstd` for those who don't know anything about it.

Reply:
Maybe even a link to the repo: https://github.com/facebook/zstd